### ISSUE TITLE
RecycleView steals data from neighbouring RecycleView

### ISSUE THREAD

**Issue by KeikiHekili** [kivy/kivy#5913]
### Versions

* Python: 2.7.14
* OS: OS X 10.11.6
* Kivy: 1.10.1
* Kivy installation method: pip

### Description

I have one RecycleView (RV1) above another (RV2), where based on the size of the elements, the RecycleView can't. I've noticed that if RV1 is scrolled to the bottom, and then RV2 is scrolled, RV2 is showing the elements that should be exclusive to RV1. The same is true if RV2 is scrolled to the bottom, and then RV1 is scrolled. In this case, RV1 shows RV2 data. In the below example, the data from RV1 is `['chocolate', 'vanilla', 'strawberry']` and the data for RV2 is `['apple', 'orange', 'banana']`. Scroll RV1 to the bottom and then start scrolling RV2, and bizarrely, RV2 will show `'apple'` and then `'vanilla'`, and then `'banana'`. 

### Code and Logs

```python
from kivy.base import runTouchApp
from kivy.lang import Builder

kv = '''
<Card@FloatLayout>:
    size_hint: None, None
    size: 250, 320
    myName: 'Chocolate'
    canvas.before:
        Color:
            rgba: 1, 1, 1, 1
        Rectangle:
            size: self.size
            pos: self.pos
    Label:
        size_hint: None, None
        size: 230, 25
        pos: self.parent.x+10, self.parent.y+295
        text: root.myName
        color: [0, 0, 0, 1]

BoxLayout:
    orientation: 'vertical'
    padding: 5
    spacing: 5
    FloatLayout:
        RecycleView:
            id: rv1
            size_hint: 25/79., 0.35
            pos_hint: {'x': 0.68, 'y': 0.5}
            viewclass: 'Card'
            data: [{'myName': i} for i in ['Chocolate', 'Vanilla', 'Strawberry']]
            RecycleBoxLayout:
                default_size: None, 320#dp(56)
                default_size_hint: 1, None
                size_hint_y: None
                height: self.minimum_height
                orientation: 'vertical'
                touch_deselect_last: False
        RecycleView:
            id: rv2
            size_hint: 25/79., 0.35
            pos_hint: {'x': 0.68, 'top': 0.495}
            viewclass: 'Card'
            data: [{'myName': i} for i in ['Apple', 'Orange', 'Banana']]
            RecycleBoxLayout:
                default_size: None, 320#dp(56)
                default_size_hint: 1, None
                size_hint_y: None
                height: self.minimum_height
                orientation: 'vertical'
                touch_deselect_last: False
'''
if __name__ == '__main__':
    runTouchApp(Builder.load_string(kv))



```

**Comment by matham** [comment]
This is the same issue as in https://github.com/kivy/kivy/issues/4666 - thanks for the runable example which helped me confirm it and figure it out.

It's quite an obvious bug in hindsight - a rookie bug actually. Not sure how I originally missed this...  😞 

This fix is simple, add
```py
    def __init__(self, **kwargs):
        super(RecycleDataAdapter, self).__init__(**kwargs)
        self.views = {}
        self.dirty_views = defaultdict(dict)
```

to `RecycleDataAdapter` in `uix/recycleview/views.py`. If someone wants to submit a PR and add a test case for this (using multiple RVs), this should be easy to do.

**Comment by KeikiHekili** [comment]
I did notice some more strange behaviour (I've edited the code into the original post). If I add a default value to the ``Card`` widgets, that value can get overridden. So, in the example, the ``someData`` value for Chocolate is initially the default value of 0, but if you scroll RV1 to the bottom, and then to the top, the ``someData`` value gets set to 2, the value from Strawberry.

**Comment by matham** [comment]
Does that happen even with the fix I mentioned?

**Comment by KeikiHekili** [comment]
Yes. I noticed this issue, and then added the fix, but this new issue still occurred.

**Comment by matham** [comment]
Yes, that's expected. If a property of a view is set through `data`, then all items must set that property. Defaults don't work the way you expect here because a view is shell for the data with no persistent state between its usage for different items.

Does that make sense?

**Comment by KeikiHekili** [comment]
Yeah, that makes sense. As you said, a matter of it working differently than I expected.

**Comment by Julian-O** [comment]
I can see that that the code has been fixed, but the documentation needs updating... but to say what? I can't see why this isn't closed.
